tests/cases/compiler/circularGetAccessor.ts

```diff
@@ -0,0 +1,3 @@
+declare class C {
+    get foo(): typeof this.foo;
+}
```

Comment:
Maybe run this test with and without `noImplicitAny`. Also, should we add a `set` accessor and rename the test `circularAccessors`?

```ts
declare class C {
    get foo(): typeof this.foo;
    set foo(value: typeof this.foo);

    get bar(): typeof this.bar;

    set baz(value: typeof this.baz);
}
```

Reply:
Not sure about the `set` accessor, it seems this is more related to checking the signature. Should TS report a circularity error for the following case?

```ts
declare class C {
    foo(a: typeof this.foo, b: typeof this.foo): void;
}
```

Reply:
I also didn't understand what error message(s) we expect on set accessors. If we have `set baz(value: typeof this.baz)`, we get an error: `Parameter 'value' of accessor has or is using private name 'this'.(4106)`
If we have something like `set foo(value: typeof foo);`, we get both `Cannot find name 'foo'.(2304)` and `Parameter 'value' of accessor has or is using private name 'foo'.(4106)`...

@DanielRosenwasser could you explain why we have the `'foo' is referenced directly or indirectly in its own type annotation.(2502)` error and when we should expect it? e.g. as a-tarasyuk commented above, should we have it when the circularity occurs in a parameter's type? should we have this error for any method, not just accessors?

Reply:
We need to come up with *some* type, and if you end up writing something like

```ts
var x: { prop: typeof x.prop };
```

then you basically need to come up with some type. We could give you `any` or `unknown` or `never` silently, but we assume that you've written some sort of circular code that might warrant an error message.

Because `getters` should (in theory!) feel indistinguishable from properties, we should provide an error on accessors. Maybe @ahejlsberg has a better handle on this.

Reply:
Ok, so coming back to this and doing some thinking, here's my understanding of it:
1. We want getters/setters to be consistent with regular properties. When you self-reference the type of the property, like `var x: { prop: typeof x.prop };`, you get the error `
'prop' is referenced directly or indirectly in its own type annotation.(2502)`, regardless of `noImplicitAny` being set or not.
2. Currently, if you do the same self-referencing in a **getter**, you only get an error if `noImplicitAny` is true. The error is caused because we guess the type is `any`, and we give a more elaborate error for this implicit any, saying that `'prop' implicitly has return type 'any' because it does not have a return type annotation and is referenced directly or indirectly in one of its return expressions.(7023)`.
This current error seems (1) a bit confusing since it does have a return type annotation, (2) inconsistent with the error we use for regular properties, and even more so because the property error happens regardless of `noImplicitAny`.

So I think this change is the right one, the new error message is to me more clear and would more quickly lead to a fix from the user.

Thinking about *setters*, though, it is not immediately clear to me we want to do the same.
1. Using a self-referential type in the type of the setter argument is sometimes ok:
```ts
declare var z: {
    get prop1(): number;
    set prop1(arg: typeof y.prop1); // Ok, `arg: number`
}
```
2. Sometimes it's not, and it seems like an implicit any:
```ts
declare var y: {
    set prop2(arg: typeof y.prop2);
};
```
(which we don't error on currently, not sure why, might be a bug).

Reply:
Ok, on the setter situation: I think it's a bit more complex than the getter case, so I'll open a separate issue and review this PR as is